### PR TITLE
Generic lease stacking

### PR DESCRIPTION
**Pull request by jakecoll**
Extends previous gpu lease stacking module to all nodes by node type. Rather than looking at individual nodes, it searches for stacked leases by node type to prevent "stair stacking" across nodes.

### CODE REVIEW

File: hammers/scripts/lease_stack_reaper.py [hunk 1]
```diff
@@ -33,25 +33,28 @@ def __init__(self, user_id, name, email):
...
         """Add lease to node list."""
-        if node_id not in self.nodes:
-            self.nodes[node_id] = []
+        if node_type not in self.nodes:
+            self.nodes[node_type] = []
 
-        self.nodes[node_id].append((lease_id, start_date, end_date))
+        # Data is by node id so do not add lease if it has already
+        # appeared for another node (i.e. multi-node leases).
+        if lease_id not in self.nodes[node_type]:
```
Comment (diurnalist) [inline hammers/scripts/lease_stack_reaper.py]: I think if `self.nodes[node_type]` was a `set()` you wouldn't need this. I would expect Python can correctly identify duplicates on tuples of basic types without having to add any sort of equality operator.

Comment (diurnalist) [reply hammers/scripts/lease_stack_reaper.py]: Also, I don't think this check will work given that you're only looking for `lease_id`.

```python
arr = [('a', 'b', 'c')]
print('a' in arr) # False
```

File: hammers/scripts/lease_stack_reaper.py [hunk 2]
```diff
@@ -33,25 +33,28 @@ def __init__(self, user_id, name, email):
...
+            self.nodes[node_type].append((lease_id, start_date, end_date))
 
     def sort_leases_by_date(self):
         """Sort leases by date for each node."""
-        for node_id, leases in self.nodes.items():
-            self.nodes[node_id] = list(sorted(
+        for node_type, leases in self.nodes.items():
+            self.nodes[node_type] = list(sorted(
```
Comment (diurnalist) [inline hammers/scripts/lease_stack_reaper.py]: Ah, looks like you are converting to a `set` here.

File: hammers/scripts/lease_stack_reaper.py [hunk 3]
```diff
@@ -75,8 +78,8 @@ def find_stacked_leases(self, leases):
             else:
                 next_start_date = datetime.max
 
-            stacked_previous = (start_date - last_end_date).days < 1
-            stacked_next = (next_start_date - end_date).days < 1
+            stacked_previous = (start_date - last_end_date).days < 2
```
Comment (diurnalist) [inline hammers/scripts/lease_stack_reaper.py]: maybe pull this "magic number" out to a constant `MIN_ALLOWED_STACK_INTERVAL_DAYS`.

File: hammers/scripts/lease_stack_reaper.py [hunk 4]
```diff
@@ -75,8 +78,8 @@ def find_stacked_leases(self, leases):
...
+            stacked_next = (next_start_date - end_date).days < 2
 
             if stacked_previous or stacked_next:
                 stacked.append(leases[i])
```
Comment (diurnalist) [inline hammers/scripts/lease_stack_reaper.py]: You could make your life easier if this included the `node_type`

```python
stacked.append((node_type, leases[i][0]))
```

You only care about the `node_type` and `lease_id` from what I can tell. The other option would be having a thin object `Lease(id=, start_date=, end_date=, node_type=)` and just passing instances of that around. Sometimes a more explicit structure helps.

File: hammers/scripts/lease_stack_reaper.py (hunk 1 above)
Comment (jakecoll) [reply hammers/scripts/lease_stack_reaper.py]: You right. Set is actually called elsewhere on leases so nothing looked unusual.

File: hammers/scripts/lease_stack_reaper.py (hunk 4 above)
Comment (jakecoll) [reply hammers/scripts/lease_stack_reaper.py]: I added in the node_type object in the print to help verify. It's not actually necessary unless we want to reference it in the email.

**Pull request merged**